### ISSUE TITLE
IPC on Windows still unavailable in v20.0

### ISSUE THREAD

**Issue by cheind** [zeromq/pyzmq#1462]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description

With pyzmq 20.0 linking against libzmq 4.3.3, `icp://` is expected to work in Windows 10 when `AF_UNIX` is available. However,  `zmq.error.ZMQError: Protocol not supported` is raised.

# Environment

* pyzmq 20.0.0
* libzmq 4.3.3
* python 3.8.6 (and possibly others)
* OS: Windows 10, Version10.0.19041 Build 19041

# Minimal test code / Steps to reproduce the issue

Ensure AF_UNIX is available `sc query afunix` then

```python
import zmq
context = zmq.Context()
socket = context.socket(zmq.PUB)
socket.bind('ipc:///tmp/stream')
```

# What's the actual result? (include assertion message & call stack if applicable)
`zmq.error.ZMQError: Protocol not supported`

# What's the expected result?
bound socket.

**Comment by cheind** [comment]
@minrk I don't think that I can help with it. I'm merely a user of pyzmq.

**Comment by minrk** [comment]
The main disadvantage of pyzmq's builds is that it doesn't (can't) run the cmake configuration to detect capabilities. It's entirely possible all we need is a `define_macros` to indicate that unix sockets should be considered available on windows.

However, I believe this also means that a build of pyzmq would have to be made on Windows 10 *and* have to require Windows 10, since it would be incompatible with earlier versions due to missing symbols (at least that's my understanding based on posix things, I'm much less familiar with Windows linking).

**Comment by minrk** [comment]
A build of libzmq from source on your machine followed by a build of pyzmq linking that libzmq might work as-is.

**Comment by cheind** [comment]
Building pyzmq myself is certainly an option (at least for me). However, I believe zmq/IPC support on Windows is a long awaited feature and it could outperform `TCP` via localhost on Windows because of the support of Posix sockets (see #1461). Hence I would strongly suggest to provide some kind of package for everyday users that supports IPC/Windows.

Could pyzmq determine support for AF_UNIX and then dynamically load the corresponding libzmq? This would probably mean that pyzmq ships with two different libzmq builds on Windows.

**Comment by cheind** [comment]
@minrk would you mind letting me know how the current release on pypi is build for windows?

**Comment by minrk** [comment]
> Could pyzmq determine support for AF_UNIX and then dynamically load the corresponding libzmq? 

It is loaded dynamically, but I'd need to understand more about how to detect AF_UNIX and how to build libzmq with this support. I don't think shipping one wheel with two complete builds of libzmq is something we should do, it should be different wheels with different capabilities, and the newer wheel should be marked with appropriate tags to refuse installation on older Windows (I don't know if Python has defined these, in which case it would need to be build-from-source-only feature).

wheels are built on CI [here](https://github.com/zeromq/pyzmq/blob/HEAD/.github/workflows/wheels.yml) with lots of complicated and fragile logic accumulated over the years in the `buildutils` directory and setup.py for how libzmq itself is built as an Extension.

**Comment by cheind** [comment]
I understand. The `build-from-source-only` strategy should be last resort IMHO, as compiling packages on Windows more tedious than on Linux for the everyday user.

**Comment by minrk** [comment]
I agree! But the consequence of having binaries means we can only ship least-common-denominator versions as binaries. It's only possible to ship a binary that only works on Windows 10 if

a) the feature fails in a reasonable way (i.e. catchable error) on older Windows, or
b) the wheel can be gated by the installer (Python on mac/linux have good-ish tags for OS feature support, I'm not sure about Windows), so that it won't get installed where it won't work.

a) would be ideal, and I would *love* help figuring out what's possible there. If a libzmq built with AF_UNIX support loads and only ipc fails on older Windows, that would be perfect. To test this, though, we are going to need to make a build of libzmq with this feature support, and run a test on an old Windows to verify how/if it fails.

**Comment by minrk** [comment]
Based on [this discussion](https://discuss.python.org/t/specifying-supported-windows-in-wheels/6712/2) I'm optimistic that with the right combination of toolchain and configuration that we will be able to ship a wheel that works with this feature enabled. It's just a pain to test because normal CI won't work (the build system and the test system *must* be different for it to be a valid test). I've lost too much time this month to debugging Windows issues in the latest wheels, so I can't put in any more time right now, but if you wanted to have a look, that would be terrific. The latest Windows wheels no longer build libzmq as part of pyzmq. Instead, they get official libzmq releases from [here](https://dl.bintray.com/zeromq/generic/). It's *possible* that just picking up the latest build there will be enough, and we only need to test that the latest build there works on the oldest supported Windows (now 8.1).

**Comment by AntiTenzor** [comment]
I very hope that IPC protocol will outperform TCP on Windows in order(s).
So could you please upgrade **pyzmq** to use latest released **libzmq** version **4.3.4** ?
I think this will switch on IPC support **automatically** .
https://github.com/zeromq/libzmq/releases/tag/v4.3.4

We all be very happy to check this long-long-waited IPC support in brilliant ZMQ. =)

**Comment by minrk** [comment]
pyzmq 22 ships with the vc141 builds of libzmq 4.3.4

**Comment by AntiTenzor** [comment]
> pyzmq 22 ships with the vc141 builds of libzmq 4.3.4

So I am completely confused what's going on.
I used command      **pip install zmq -U**
It gave me **PyZMQ v18.1.1**   binded to **libzmq v4.3.2**

Should I completely purge outdated package **zmq** in favor of **pyzmq**?

**Comment by AntiTenzor** [comment]
Btw, the IPC protocol doesn't work on Windows Server 2019 Standard 64 bit.
Python v3.7.9    64bit.

Here is my code:
`
import zmq
import platform

url = 'ipc://zmq_pub_54321'

print(f'Version of LibZMQ: {zmq.zmq_version()}')
print(f'Version of PyZMQ : {zmq.pyzmq_version()}')
print()

print(f'System           :   {platform.system()}')
print(f'Platform         :   {platform.platform()}')
print(f'Architecture     :   {platform.architecture()}')
print()



print('Creating the CONTEXT...')
zmqContext = zmq.Context()
print('Creating the PUBLISHER...')
publisher = zmqContext.socket(zmq.PUB)

print(f'Binding PUBLISHER to its address   {url}  ...')
publisher.bind(url)
`

Here is its output:
`
Version of LibZMQ: 4.3.4
Version of PyZMQ : 22.0.3

System           :   Windows
Platform         :   Windows-10-10.0.17763-SP0
Architecture     :   ('64bit', 'WindowsPE')

Creating the CONTEXT...
Creating the PUBLISHER...
Binding PUBLISHER to its address   ipc://zmq_pub_54321  ...
Traceback (most recent call last):
  File "pub.py", line 31, in <module>
    publisher.bind(url)
  File "C:\Run\Python37\lib\site-packages\zmq\sugar\socket.py", line 172, in bind
    super().bind(addr)
  File "zmq\backend\cython\socket.pyx", line 540, in zmq.backend.cython.socket.Socket.bind
  File "zmq\backend\cython\checkrc.pxd", line 28, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Protocol not supported

`

**Comment by minrk** [comment]
I'm not sure how IPC works in libzmq, so I don't know if this is an issue with the libzmq builds not having it enabled or something else.

**Comment by mattrjackson** [comment]
I got this working today hacking around the files. Turns out there are a couple of small issues keeping this from working:

1. buildutils/include_win32/platform.hpp is missing the preprocessor definition enabling IPC, which in windows evidently also requires that epoll be set as the poller:
```
#define ZMQ_IOTHREAD_POLLER_USE_EPOLL 1
/* Have AF_UNIX sockets for ipc transport */
#define ZMQ_HAVE_IPC 1
```
2. ipcmaxlen.h and check_sys_un.c are missing a check for the windows version of unix sockets (afunix.h). Copying what I found in the base libzmq, you can replace:
```
#include <sys/un.h>
```
with
```
#if defined _MSC_VER
#include <afunix.h>
#else
#include <sys/un.h>
#endif
```
3. Next we need to compile wepoll.c as well. This requires adding around line 571 something akin to:
```
sources.append('path/to/wepoll.c')
```  
It would be nice to have this fixed for the next release. I will try to put a PR together this weekend.
